assertion/function/assertiontree/backprop_util.go

```diff
@@ -307,6 +319,33 @@ func handleErrorReturns(rootNode *RootAssertionNode, retStmt *ast.ReturnStmt, re
 	}
 }
 
+// handleBooleanReturns handles the special case for boolean (`ok`) returning functions (n-th result of type `bool`
+// which guards at least one of the first n-1 non-bool results). Similar to the handliong of error returning functions,
+// for boolean returns, we generate consumers by applying the following boolean contract:
+// (1) if boolean return value = true, create consumers for the non-boolean returns
+// TODO: currently we support only explicit boolean returns (i.e., `return r0, r1, ..., {true|false}`). We should also support implicit boolean returns, i.e., `return` or `return <expr>` in the future.
+//
+// handleBooleanReturns returns true if the above contract is satisfied and consumers are created, false otherwise
+func handleBooleanReturns(rootNode *RootAssertionNode, retStmt *ast.ReturnStmt, results []ast.Expr, isNamedReturn bool) bool {
+	nRetIndex := len(results) - 1
+	nRetExpr := results[nRetIndex]          // n-th expression
+	nMinusOneRetExpr := results[:nRetIndex] // n-1 expressions
+
+	// check if the return statement is of the currently supported explicit boolean return form (`return ..., {true|false}`)
+	val, ok := nRetExpr.(*ast.Ident)
+	if !ok || (val.Name != "true" && val.Name != "false") {
+		return false
+	}
```

Comment:
I remember we have a `util.IsLiteral` helper function to check if an expression is a literal nil, true, or false. Right? Let's use that :) 

Another thought, for a follow-up PR, is to enhance the `IsLiteral` function: currently, we are checking if the AST node is an identifier and the name is either `true` or `false`. This is error-prone (although unlikely to happen in real production code): devs can 

(1) shadow `true` as a variable:
```go
func foo() bool {
  var true bool = false
  return true // whoops, nilaway thinks this is true
}
```

(2) return a compile-time constant:

```go
const aa = true

func foo() bool {
  return aa // nilaway doesn't think this is returning true.
}
```

We should leverage proper type-checking and use `pass.TypesInfo.Type[expr]` to get a https://pkg.go.dev/go/types#TypeAndValue, and then use https://pkg.go.dev/go/constant#Val to check its compile-time value. We could probably also put it in a helper function in, say `typeshelper` :)

Reply:
You bring up an interesting point! Although rare, you are right that we should account for shadowed built-ins and compile time constants. (2) is done, since it was fairly straightforward to implement. (1) needs more work, which I'll try to get in this PR, or maybe a follow-up. However, the strengthening of the `isLiteral` will definitely be done in a follow-up, since it is used at other places unrelated to this PR as well.